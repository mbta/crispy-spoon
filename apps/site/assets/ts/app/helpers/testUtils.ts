/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
