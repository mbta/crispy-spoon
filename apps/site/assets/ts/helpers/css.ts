import {
  isABusRoute,
  isACommuterRailRoute,
  isAGreenLineRoute,
  isASilverLineRoute
} from "../models/route";
import { Route } from "../__v3api";

export const routeToCSSClass = (route: string): string =>
  route.toLowerCase().replace(/[ _]/g, "-");

export const busClass = (route: Route): string =>
  isABusRoute(route) && !isASilverLineRoute(route) ? "bus-route-sign" : "";

export const routeToModeName = (route: Route): string => {
  if (isACommuterRailRoute(route)) return "commuter-rail";
  if (route.type === 4) return "ferry";
  if (route.id === "Red" || route.id === "Mattapan") return "red-line";
  if (route.id === "Orange") return "orange-line";
  if (route.id === "Blue") return "blue-line";
  if (isAGreenLineRoute(route)) return "green-line";
  if (isASilverLineRoute(route)) return "silver-line";
  if (isABusRoute(route)) return "bus";
  return "unknown";
};

export const routeBgClass = (route: Route): string =>
  `u-bg--${routeToModeName(route)}`;
