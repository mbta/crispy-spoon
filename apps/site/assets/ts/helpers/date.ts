// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a date in the form of "Aug 7"
export const shortDate = (date: Date): string =>
  date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    timeZone: "UTC"
  });

export default formattedDate;
