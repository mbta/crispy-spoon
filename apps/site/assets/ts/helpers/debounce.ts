// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export const debouncePromise = (
  callback: Function,
  delay: number
): Function => {
  let timerId: number | undefined;

  return function debounced(...args: unknown[]) {
    if (timerId) {
      window.clearTimeout(timerId);
    }

    return new Promise(resolve => {
      timerId = window.setTimeout(() => resolve(callback(...args)), delay);
    });
  };
};

export default debounce;
