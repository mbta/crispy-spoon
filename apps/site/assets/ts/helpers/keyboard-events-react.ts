import { KeyboardEvent as ReactKeyboardEvent } from "react";

// for React
export const handleReactEnterKeyPress = (
  e: ReactKeyboardEvent,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const handleReactExitKeyPress = (
  e: ReactKeyboardEvent,
  onClick: Function
): void => {
  if (e.key === "Escape") {
    onClick();
  }
};
