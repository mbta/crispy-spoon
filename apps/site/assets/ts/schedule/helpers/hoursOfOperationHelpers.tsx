import React, { ReactElement } from "react";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "../components/__schedule";

const pdfLink = (pdf: SchedulePDF | null): ReactElement<HTMLElement> | null => {
  if (!pdf) {
    return null;
  }
  return (
    <div className="fs-16 pt-16">
      <a
        key={pdf.url}
        href={pdf.url}
        rel="noopener noreferrer"
        className="c-call-to-action"
        target="_blank"
      >
        {`Open full schedule and map PDF`}{" "}
        {renderFa("fa-arrow-up-right-from-square", "")}
      </a>
    </div>
  );
};

export default pdfLink;
