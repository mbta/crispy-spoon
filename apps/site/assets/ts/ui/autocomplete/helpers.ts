import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
