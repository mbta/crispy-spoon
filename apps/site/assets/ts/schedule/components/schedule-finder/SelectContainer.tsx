import React, { ReactElement, ReactNode } from "react";
import { handleReactEnterKeyPress } from "../../../helpers/keyboard-events-react";
import renderSvg from "../../../helpers/render-svg";
import arrowIcon from "../../../../static/images/icon-down-arrow.svg";

interface SelectContainerProps {
  children: ReactNode;
  error?: boolean;
  handleClick?: Function;
}

const SelectContainer = ({
  children,
  error = false,
  handleClick
}: SelectContainerProps): ReactElement<HTMLElement> => (
  <div
    tabIndex={0}
    className={`c-select-custom__container ${error ? "error" : ""}`}
    role="button"
    onClick={e => {
      if (handleClick) {
        handleClick(e);
      }
    }}
    onKeyUp={e =>
      handleReactEnterKeyPress(e, () => {
        if (handleClick) {
          handleClick(e);
        }
      })
    }
  >
    {children}
    {renderSvg("c-svg__icon c-select-custom__arrow", arrowIcon, true)}
  </div>
);

export default SelectContainer;
