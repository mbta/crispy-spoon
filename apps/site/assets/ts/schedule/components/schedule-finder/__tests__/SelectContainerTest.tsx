import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import * as KeyboardEvents from "../../../../helpers/keyboard-events-react";
import SelectContainer from "../SelectContainer";

jest.mock("../../../../helpers/keyboard-events-react", () => ({
  __esModule: true,
  handleReactEnterKeyPress: jest.fn()
}));

describe("SelectContainer", () => {
  it("renders", () => {
    const tree = renderer.create(
      <SelectContainer>
        <div>Children</div>
      </SelectContainer>
    );

    expect(tree).toMatchSnapshot();
  });

  it("highlights an error", () => {
    const wrapper = shallow(
      <SelectContainer error={true}>
        <div>Children</div>
      </SelectContainer>
    );

    expect(
      wrapper.find(".c-select-custom__container.error").exists
    ).toBeTruthy();
  });

  it("handles clicks", () => {
    const handleClick = jest.fn();

    const wrapper = shallow(
      <SelectContainer handleClick={handleClick}>
        <div>Children</div>
      </SelectContainer>
    );

    wrapper.find(".c-select-custom__container").simulate("click");

    expect(handleClick).toHaveBeenCalled();
  });

  it("accepts the Enter key as the same as a click", () => {
    const mockHandleReactEnterKeyPress: jest.Mock = KeyboardEvents.handleReactEnterKeyPress as jest.Mock;
    mockHandleReactEnterKeyPress.mockImplementation((_, func) => {
      func();
    });
    const handleClick = jest.fn();

    const wrapper = shallow(
      <SelectContainer handleClick={handleClick}>
        <div>Children</div>
      </SelectContainer>
    );

    wrapper
      .find(".c-select-custom__container")
      .simulate("keyUp", { key: "Enter" });

    expect(mockHandleReactEnterKeyPress).toHaveBeenCalled();
    expect(handleClick).toHaveBeenCalled();
  });
});
