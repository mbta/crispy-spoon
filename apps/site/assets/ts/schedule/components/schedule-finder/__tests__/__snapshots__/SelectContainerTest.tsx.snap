// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectContainer renders 1`] = `
<div
  className="c-select-custom__container "
  onClick={[Function]}
  onKeyUp={[Function]}
  role="button"
  tabIndex={0}
>
  <div>
    Children
  </div>
  <span
    aria-hidden="true"
    className="notranslate c-svg__icon c-select-custom__arrow"
    dangerouslySetInnerHTML={
      Object {
        "__html": "SVG",
      }
    }
  />
</div>
`;
